commit 1883b0e9f53ebe59b750219d8313e7049c761858
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:41:26 2020 +0100

    1010-fix-retransmit-intervals.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 71275e4..44fe7e4 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -3625,21 +3625,21 @@ static void show_ip_ospf_interface_sub(struct vty *vty, struct ospf *ospf,
 
 		if (use_json) {
 			if (OSPF_IF_PARAM(oi, fast_hello) == 0)
-				json_object_int_add(
-					json_interface_sub, "timerMsecs",
-					OSPF_IF_PARAM(oi, v_hello) * 1000);
+				json_object_int_add(json_interface_sub,
+						    "timerMsecs",
+						    OSPF_IF_PARAM(oi, v_hello));
 			else
 				json_object_int_add(
 					json_interface_sub, "timerMsecs",
-					1000 / OSPF_IF_PARAM(oi, fast_hello));
+					OSPF_IF_PARAM(oi, fast_hello));
 			json_object_int_add(json_interface_sub,
-					    "timerDeadSecs",
+					    "timerDeadMSecs",
 					    OSPF_IF_PARAM(oi, v_wait));
 			json_object_int_add(json_interface_sub,
-					    "timerWaitSecs",
+					    "timerWaitMSecs",
 					    OSPF_IF_PARAM(oi, v_wait));
 			json_object_int_add(
-				json_interface_sub, "timerRetransmitSecs",
+				json_interface_sub, "timerRetransmit",
 				OSPF_IF_PARAM(oi, retransmit_interval));
 		} else {
 			vty_out(vty, "  Timer intervals configured,");
