commit 9f758b77e31572eea202e33178351052a421d4a2
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:08 2020 +0100

    1020-fix-neighbor-status.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 45adf53..60a7068 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -4952,6 +4952,23 @@ static void show_ip_ospf_neighbor_detail_sub(struct vty *vty,
 		vty_out(vty, "    In the area %s via interface %s\n",
 			ospf_area_desc_string(oi->area), oi->ifp->name);
 
+	/* Show Transit Area ID */
+	if (use_json) {
+		if (memcmp(oi->ifp->name, "VLINK", 5) == 0) {
+			if (oi->vl_data != NULL) {
+				json_object_string_add(
+					json_neigh, "transitAreaId",
+					inet_ntoa(oi->vl_data->vl_area_id));
+			} else {
+				json_object_string_add(
+					json_neigh, "transitAreaId", "Unknown");
+			}
+		} else {
+			json_object_string_add(json_neigh, "transitAreaId",
+					       ospf_area_desc_string(oi->area));
+		}
+	}
+
 	/* Show neighbor priority and state. */
 	if (use_json) {
 		json_object_int_add(json_neigh, "nbrPriority", nbr->priority);
