commit f7b9c9e2b1b9a90ec2d15b38101bb0c097cd077d
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:09 2020 +0100

    1040-as-external-cost.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 60a7068..d8cceba 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9297,6 +9297,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 						       "N E1");
 				json_object_int_add(json_route, "cost",
 						    er->cost);
+				json_object_int_add(json_route, "ext_cost",
+						    er->u.ext.type2_cost);
 			} else {
 				vty_out(vty,
 					"N E1 %-18s    [%d] tag: %" ROUTE_TAG_PRI
@@ -9310,6 +9312,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 						       "N E2");
 				json_object_int_add(json_route, "cost",
 						    er->cost);
+				json_object_int_add(json_route, "ext_cost",
+						    er->u.ext.type2_cost);
 			} else {
 				vty_out(vty,
 					"N E2 %-18s    [%d/%d] tag: %" ROUTE_TAG_PRI
