commit e59d0285a105e56f6630b483eee336b80216e47f
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:09 2020 +0100

    1050-fix-ospf-route-status.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index d8cceba..73b7197 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9141,11 +9141,13 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 		if (rn->info == NULL)
 			continue;
 		int flag = 0;
+		char buf1[19];
 
 		json_route = json_object_new_object();
 		if (json) {
-			json_object_object_add(json, inet_ntoa(rn->p.u.prefix4),
-					       json_route);
+			snprintf(buf1, 19, "%s/32", inet_ntoa(rn->p.u.prefix4));
+
+			json_object_object_add(json, buf1, json_route);
 			json_object_string_add(json_route, "routeType", "R ");
 		} else {
 			vty_out(vty, "R    %-15s    ",
