commit a7d0fa5b9e2a5e56ef26118cc098250485358c9c
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:11 2020 +0100

    1080-fix-ospf-route-status-multiple-areas.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index f6f8135..b6fb304 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9442,8 +9442,8 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 	struct ospf_route * or ;
 	struct listnode *pnode, *pnnode;
 	struct ospf_path *path;
-	json_object *json_route = NULL, *json_nexthop_array = NULL,
-		    *json_nexthop = NULL;
+	json_object *json_route_array = NULL, *json_route = NULL,
+		    *json_nexthop_array = NULL, *json_nexthop = NULL;
 
 	if (!json)
 		vty_out(vty,
@@ -9457,9 +9457,10 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 		memset(buf1, 0, sizeof(buf1));
 		prefix2str(&rn->p, buf1, sizeof(buf1));
 
-		json_route = json_object_new_object();
 		if (json) {
-			json_object_object_add(json, buf1, json_route);
+			json_route_array = json_object_new_array();
+			json_route = json_object_new_object();
+			json_object_object_add(json, buf1, json_route_array);
 			json_object_to_json_string_ext(
 				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
@@ -9577,8 +9578,8 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 				}
 			}
 		}
-		if (!json)
-			json_object_free(json_route);
+		if (json)
+			json_object_array_add(json_route_array, json_route);
 	}
 	if (!json)
 		vty_out(vty, "\n");
@@ -9593,8 +9594,8 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 	struct listnode *pnode;
 	struct listnode *node;
 	struct ospf_path *path;
-	json_object *json_route = NULL, *json_nexthop_array = NULL,
-		    *json_nexthop = NULL;
+	json_object *json_route_array = NULL, *json_route = NULL,
+		    *json_nexthop_array = NULL, *json_nexthop = NULL;
 
 	if (!json)
 		vty_out(vty,
@@ -9606,12 +9607,12 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 		int flag = 0;
 		char buf1[19];
 
-		json_route = json_object_new_object();
 		if (json) {
+			json_route = json_object_new_object();
 			snprintf(buf1, 19, "%s/32", inet_ntoa(rn->p.u.prefix4));
 
-			json_object_object_add(json, buf1, json_route);
-			json_object_string_add(json_route, "routeType", "R ");
+			json_object_object_add(json, buf1, json_route_array);
+
 		} else {
 			vty_out(vty, "R    %-15s    ",
 				inet_ntoa(rn->p.u.prefix4));
@@ -9625,6 +9626,9 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 
 			/* Show path. */
 			if (json) {
+				json_route = json_object_new_object();
+				json_object_string_add(json_route, "routeType",
+						       "R ");
 				json_object_int_add(json_route, "cost",
 						    or->cost);
 				json_object_string_add(
@@ -9717,9 +9721,10 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 					}
 				}
 			}
+			if (json)
+				json_object_array_add(json_route_array,
+						      json_route);
 		}
-		if (!json)
-			json_object_free(json_route);
 	}
 	if (!json)
 		vty_out(vty, "\n");
@@ -9733,8 +9738,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 	struct ospf_route *er;
 	struct listnode *pnode, *pnnode;
 	struct ospf_path *path;
-	json_object *json_route = NULL, *json_nexthop_array = NULL,
-		    *json_nexthop = NULL;
+	json_object *json_route_array = NULL, *json_route = NULL,
+		    *json_nexthop_array = NULL, *json_nexthop = NULL;
 
 	if (!json)
 		vty_out(vty,
@@ -9748,9 +9753,11 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 
 		snprintf(buf1, sizeof(buf1), "%s/%d",
 			 inet_ntoa(rn->p.u.prefix4), rn->p.prefixlen);
-		json_route = json_object_new_object();
+
 		if (json) {
-			json_object_object_add(json, buf1, json_route);
+			json_route_array = json_object_new_array();
+			json_route = json_object_new_object();
+			json_object_object_add(json, buf1, json_route_array);
 			json_object_to_json_string_ext(
 				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
@@ -9846,8 +9853,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 				}
 			}
 		}
-		if (!json)
-			json_object_free(json_route);
+		if (json)
+			json_object_array_add(json_route_array, json_route);
 	}
 	if (!json)
 		vty_out(vty, "\n");
