commit f8273ce0f870b1486e7a010b940c76bc38258e7c
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:12 2020 +0100

    1110-fix-ospf-route-status-performance.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 96a09c8..baedd57 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9461,8 +9461,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (or->path_type) {
@@ -9768,8 +9766,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (er->path_type) {
