commit 4a88d24f3813541f4c530e54554e502e004d6935
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:16 2020 +0100

    1230-ospfv3-fix-interface_router_id_empty_fix.patch

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index b91b03e..de72701 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -1000,6 +1000,9 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp,
 					json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf,
+				  sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
