commit 159c23e71acfa003606a95f9ad78acfd91eda1e0
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 29 10:30:17 2020 +0100

    1320-fix-rtm_getrule-log-error.patch

diff --git a/zebra/kernel_netlink.c b/zebra/kernel_netlink.c
index a1f7014..16e6c87 100644
--- a/zebra/kernel_netlink.c
+++ b/zebra/kernel_netlink.c
@@ -905,7 +905,8 @@ int netlink_parse_info(int (*filter)(struct nlmsghdr *, ns_id_t, int),
 							err->msg.nlmsg_seq,
 							err->msg.nlmsg_pid);
 				} else {
-					if ((msg_type != RTM_GETNEXTHOP)
+					if ((msg_type != RTM_GETNEXTHOP
+					     && msg_type != RTM_GETRULE)
 					    || !startup)
 						flog_err(
 							EC_ZEBRA_UNEXPECTED_MESSAGE,
