commit ef8fa437b6c0e9cb5cdc49f7cbf831396a0f55ff
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:50 2020 +0100

    1010-fix-retransmit-intervals.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 0ac9d4a..14c6870 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -3594,19 +3594,19 @@ static void show_ip_ospf_interface_sub(struct vty *vty, struct ospf *ospf,
 			if (OSPF_IF_PARAM(oi, fast_hello) == 0)
 				json_object_int_add(
 					json_interface_sub, "timerMsecs",
-					OSPF_IF_PARAM(oi, v_hello) * 1000);
+					OSPF_IF_PARAM(oi, v_hello));
 			else
 				json_object_int_add(
 					json_interface_sub, "timerMsecs",
-					1000 / OSPF_IF_PARAM(oi, fast_hello));
+					OSPF_IF_PARAM(oi, fast_hello));
 			json_object_int_add(json_interface_sub,
-					    "timerDeadSecs",
+					    "timerDeadMSecs",
 					    OSPF_IF_PARAM(oi, v_wait));
 			json_object_int_add(json_interface_sub,
-					    "timerWaitSecs",
+					    "timerWaitMSecs",
 					    OSPF_IF_PARAM(oi, v_wait));
 			json_object_int_add(
-				json_interface_sub, "timerRetransmitSecs",
+				json_interface_sub, "timerRetransmit",
 				OSPF_IF_PARAM(oi, retransmit_interval));
 		} else {
 			vty_out(vty, "  Timer intervals configured,");
