commit 38d92bd8af3c0c5db308f0552d5661019c754784
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:50 2020 +0100

    1050-fix-ospf-route-status.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index da365b7..a617144 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9103,10 +9103,13 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 		if (rn->info == NULL)
 			continue;
 		int flag = 0;
+		char buf1[19];
 
 		json_route = json_object_new_object();
 		if (json) {
-			json_object_object_add(json, inet_ntoa(rn->p.u.prefix4),
+			snprintf(buf1, 19, "%s/32", inet_ntoa(rn->p.u.prefix4));
+
+			json_object_object_add(json, buf1,
 					       json_route);
 			json_object_string_add(json_route, "routeType", "R ");
 		} else {
