commit 2c44cbe3f0a5ca58089abae6ef779f9dd76cc714
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:51 2020 +0100

    1110-fix-ospf-route-status-performance.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 3422cc6..47f294c 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9296,8 +9296,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (or->path_type) {
@@ -9594,8 +9592,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (er->path_type) {
