commit 737eae050fe31a28328a1c56afbb4163c2d12424
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:51 2020 +0100

    1150-fix-rip-clean-do-peer-clean.patch

diff --git a/ripd/rip_peer.c b/ripd/rip_peer.c
index ccd3273..7408199 100644
--- a/ripd/rip_peer.c
+++ b/ripd/rip_peer.c
@@ -107,6 +107,24 @@ static struct rip_peer *rip_peer_get(struct in_addr *addr)
 	return peer;
 }
 
+void rip_peer_clean(void)
+{
+	struct rip_peer *peer;
+	struct listnode *node, *nnode;
+
+	for (ALL_LIST_ELEMENTS(peer_list, node, nnode, peer)) {
+		if (peer) {
+			if (peer->t_timeout) {
+				thread_cancel(peer->t_timeout);
+			}
+			listnode_delete(peer_list, peer);
+			rip_peer_free(peer);
+		}
+	}
+	return;
+}
+
+
 void rip_peer_update(struct sockaddr_in *from, uint8_t version)
 {
 	struct rip_peer *peer;
diff --git a/ripd/ripd.c b/ripd/ripd.c
index 5c53a04..4549ce9 100644
--- a/ripd/ripd.c
+++ b/ripd/ripd.c
@@ -4180,6 +4180,7 @@ void rip_clean(void)
 	rip_interfaces_clean();
 	rip_distance_reset();
 	rip_redistribute_clean();
+	rip_peer_clean();
 }
 
 /* Reset all values to the default settings. */
diff --git a/ripd/ripd.h b/ripd/ripd.h
index a47e626..3b6d65a 100644
--- a/ripd/ripd.h
+++ b/ripd/ripd.h
@@ -409,6 +409,7 @@ extern void rip_peer_bad_packet(struct sockaddr_in *);
 extern void rip_peer_display(struct vty *, json_object *, uint8_t);
 extern struct rip_peer *rip_peer_lookup(struct in_addr *);
 extern struct rip_peer *rip_peer_lookup_next(struct in_addr *);
+extern void rip_peer_clean(void);
 
 extern int rip_offset_list_apply_in(struct prefix_ipv4 *, struct interface *,
 				    uint32_t *);
