commit 3a6d8feb7a9a79b7280ff81a3e7bfe07a04952d9
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:51 2020 +0100

    1160-fix-rip-show-authmode-config.patch

diff --git a/ripd/rip_interface.c b/ripd/rip_interface.c
index 6cba821..db3f838 100644
--- a/ripd/rip_interface.c
+++ b/ripd/rip_interface.c
@@ -1751,6 +1751,7 @@ static int rip_interface_config_write(struct vty *vty)
 		    && (ri->ri_send == RI_RIP_UNSPEC)
 		    && (ri->ri_receive == RI_RIP_UNSPEC)
 		    && (ri->auth_type != RIP_AUTH_MD5) && (!ri->v2_broadcast)
+		    && (ri->auth_type != RIP_AUTH_SIMPLE_PASSWORD)
 		    && (ri->md5_auth_len != RIP_AUTH_MD5_SIZE)
 		    && (!ri->auth_str) && (!ri->key_chain))
 			continue;
