commit 8bf0627275a96d4460f0e5cef7387f156bf90bf1
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:51 2020 +0100

    1230-ospfv3-fix-interface_router_id_empty_fix.patch

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index 0e25e81..afb3d48 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -955,6 +955,8 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp, uint8_t
 				json_object_boolean_false_add(json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf, sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
