commit f38e91c267a0760d603c7347956f85b71d921e40
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Nov 3 13:56:52 2020 +0100

    1270-fix-false-not-defined.patch

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 47f294c..8ebfe19 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -53,6 +53,12 @@
 #include "ospfd/ospf_dump.h"
 #include "ospfd/ospf_bfd.h"
 
+#ifndef FALSE
+// Used to be defined in json_object.h (from libjson-c), but was removed with
+// the release of json-c-0.14-20200419.
+#define FALSE 0
+#endif
+
 static const char *ospf_network_type_str[] = {
 	"Null",	"POINTOPOINT", "BROADCAST", "NBMA", "POINTOMULTIPOINT",
 	"VIRTUALLINK", "LOOPBACK"};
