diff --git a/configure.ac b/configure.ac
index ef76b5b..0c04b30 100755
--- a/configure.ac
+++ b/configure.ac
@@ -605,7 +605,13 @@ AS_IF([test "$host" = "$build"], [
 	python2.7-config ])
   if test -n "$PYTHONCONFIG"; then
     PYTHON_CFLAGS="`\"${PYTHONCONFIG}\" --includes`"
-    PYTHON_LIBS="`\"${PYTHONCONFIG}\" --ldflags`"
+
+    case "$PYTHONCONFIG" in
+      *"python3"*)
+        PYTHON_LIBS="`\"${PYTHONCONFIG}\" --ldflags --embed`" ;;
+      *)
+        PYTHON_LIBS="`\"${PYTHONCONFIG}\" --ldflags`" ;;
+    esac
 
     FRR_PYTHON_CHECK_WORKING([])
   fi
