commit aeb7f3e42f02d741e74ec52c1c3f47f2d63cd402
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:03:19 2020 +0200

    4.0-1050

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 510dce22c..89c570eab 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9222,11 +9222,14 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 		if (rn->info == NULL)
 			continue;
 		int flag = 0;
+		char buf1[19];
 
 		json_route = json_object_new_object();
 		if (json) {
+			snprintf(buf1, 19, "%s/32", inet_ntoa(rn->p.u.prefix4));
+	
 			json_object_object_add(json,
-					inet_ntoa(rn->p.u.prefix4),
+					buf1,
 					json_route);
 			json_object_string_add(json_route, "routeType",
 					       "R ");
