commit d4f57272b0c78f4216957c4185bf9a2f9b8dbeac
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:06:05 2020 +0200

    4.0-1090

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index f5bad444f..6795162d2 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9591,17 +9591,18 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 					json_object_boolean_true_add(
 							json_route,
 							"IA");
-				if (or->u.std.flags & ROUTER_LSA_BORDER)
-					json_object_string_add(
-							json_route,
-							"routerType",
-							"abr");
-				else if (or->u.std.flags &
-					 ROUTER_LSA_EXTERNAL)
-					json_object_string_add(
-							json_route,
-							"routerType",
-							"asbr");
+				if ((or->u.std.flags & (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER))
+					== (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER)) {
+					snprintf(buf1, sizeof(buf1), "abr asbr");
+				} else if (or->u.std.flags & ROUTER_LSA_BORDER) {
+					snprintf(buf1, sizeof(buf1), "abr");
+				} else if (or->u.std.flags & ROUTER_LSA_EXTERNAL) {
+					snprintf(buf1, sizeof(buf1), "asbr");
+				}
+				json_object_string_add(
+						json_route,
+						"routerType",
+						buf1);
 			} else {
 				vty_out(vty, "%s [%d] area: %s",
 				(or->path_type == OSPF_PATH_INTER_AREA
