commit 9e86982e11ea043d8b97cafaee215d85a21c98e5
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:15:43 2020 +0200

    4.0-1110

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index a21e36019..08237d56a 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9411,8 +9411,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			 json_route_array = json_object_new_array();
 			 json_route = json_object_new_object();
 			 json_object_object_add(json, buf1, json_route_array);
-			 json_object_to_json_string_ext(json,
-				JSON_C_TO_STRING_NOSLASHESCAPE);
 
 		}
 
@@ -9717,9 +9715,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			 json_route_array = json_object_new_array();
 			 json_route = json_object_new_object();
 			 json_object_object_add(json, buf1, json_route_array);
-			 json_object_to_json_string_ext(json,
-				JSON_C_TO_STRING_NOSLASHESCAPE);
-
 		}
 
 		switch (er->path_type) {
