commit 609c3f7203434806f8cb6b98b99748dd28974876
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:16:42 2020 +0200

    4.0-1140

diff --git a/ripd/ripd.c b/ripd/ripd.c
index ec888b101..8c17a7b84 100644
--- a/ripd/ripd.c
+++ b/ripd/ripd.c
@@ -2321,13 +2321,15 @@ void rip_output_process(struct connected *ifc, struct sockaddr_in *to,
 							  tmp_rinfo))
 					if (tmp_rinfo->type == ZEBRA_ROUTE_RIP
 					    && tmp_rinfo->nh.ifindex
-						       == ifc->ifp->ifindex)
-						rinfo->metric_out =
+					    == ifc->ifp->ifindex)
+						tmp_rinfo->metric_out =
 							RIP_METRIC_INFINITY;
-				if (tmp_rinfo->type == ZEBRA_ROUTE_CONNECT
+
+				if (rinfo->type == ZEBRA_ROUTE_CONNECT
 				    && prefix_match((struct prefix *)p,
 						    ifc->address))
-					rinfo->metric_out = RIP_METRIC_INFINITY;
+					rinfo->metric_out =
+						RIP_METRIC_INFINITY;
 			}
 
 			/* Prepare preamble, auth headers, if needs be */
