commit 21a05b4e648af01e2c61af822bca6bb6fbc8d024
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:17:28 2020 +0200

    4.0-1160

diff --git a/ripd/rip_interface.c b/ripd/rip_interface.c
index f6723f32e..2d7550ce3 100644
--- a/ripd/rip_interface.c
+++ b/ripd/rip_interface.c
@@ -1754,6 +1754,7 @@ static int rip_interface_config_write(struct vty *vty)
 		    && (ri->ri_send == RI_RIP_UNSPEC)
 		    && (ri->ri_receive == RI_RIP_UNSPEC)
 		    && (ri->auth_type != RIP_AUTH_MD5) && (!ri->v2_broadcast)
+		    && (ri->auth_type != RIP_AUTH_SIMPLE_PASSWORD)
 		    && (ri->md5_auth_len != RIP_AUTH_MD5_SIZE)
 		    && (!ri->auth_str) && (!ri->key_chain))
 			continue;
