commit 4e5af64270152f74707240343b5d2ee6184ea2d9
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:17:39 2020 +0200

    4.0-1170

diff --git a/configure.ac b/configure.ac
index 196830a03..f49ac4d10 100755
--- a/configure.ac
+++ b/configure.ac
@@ -1765,7 +1765,7 @@ AM_CONDITIONAL([ZEROMQ], test "x$ZEROMQ" = "xtrue")
 dnl ----------
 dnl configure date
 dnl ----------
-CONFDATE=`date '+%Y%m%d'`
+CONFDATE=0
 AC_SUBST(CONFDATE)
 
 dnl ------------------------------
