commit 533370ef01d932b9d59de1c2f1c59c44d5dad9b6
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Tue Oct 6 13:20:16 2020 +0200

    4.0-1230

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index 743c6b433..6d3185e3c 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -982,6 +982,8 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp, u_char u
 				json_object_boolean_false_add(json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf, sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
