commit 3d1b5d051288ecb7f60e0542debdbcb459ac9142
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 15:01:31 2020 +0200

    5.0-1040

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index ba0104cb1..7e352c1c0 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9292,6 +9292,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 						       "N E1");
 				json_object_int_add(json_route, "cost",
 						    er->cost);
+				json_object_int_add(json_route, "ext_cost",
+							    er->u.ext.type2_cost);
 			} else {
 				vty_out(vty,
 					"N E1 %-18s    [%d] tag: %" ROUTE_TAG_PRI
@@ -9305,6 +9307,8 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 						       "N E2");
 				json_object_int_add(json_route, "cost",
 						    er->cost);
+				json_object_int_add(json_route, "ext_cost",
+							    er->u.ext.type2_cost);
 			} else {
 				vty_out(vty,
 					"N E2 %-18s    [%d/%d] tag: %" ROUTE_TAG_PRI
