commit 4a69ad42fd195672e8aa8c57d2ce344c7a5d1771
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 15:04:03 2020 +0200

    5.0-1090

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 7d5b27387..f96879fc0 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9502,14 +9502,19 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 				if (or->path_type == OSPF_PATH_INTER_AREA)
 					json_object_boolean_true_add(json_route,
 								     "IA");
-				if (or->u.std.flags & ROUTER_LSA_BORDER)
-					json_object_string_add(json_route,
-							       "routerType",
-							       "abr");
-				else if (or->u.std.flags & ROUTER_LSA_EXTERNAL)
-					json_object_string_add(json_route,
-							       "routerType",
-							       "asbr");
+				if ((or->u.std.flags & (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER))
+					== (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER)) {
+					snprintf(buf1, sizeof(buf1), "abr asbr");
+				} else if (or->u.std.flags & ROUTER_LSA_BORDER) {
+					snprintf(buf1, sizeof(buf1), "abr");
+				} else if (or->u.std.flags & ROUTER_LSA_EXTERNAL) {
+					snprintf(buf1, sizeof(buf1), "asbr");
+				} else {
+					buf1[0] = '\0';
+				}
+				json_object_string_add(json_route,
+						       "routerType",
+						       buf1);
 			} else {
 				vty_out(vty, "%s [%d] area: %s",
 					(or->path_type == OSPF_PATH_INTER_AREA
