commit bb3f2046274e6501ab893961476ca5340d7fb5bd
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 15:04:55 2020 +0200

    5.0-1110

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 1cce3584a..1d143765e 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9329,8 +9329,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (or->path_type) {
@@ -9627,8 +9625,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (er->path_type) {
