commit 2cf911b14ff26916f3f11dad9ec8157556479a1d
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 10:27:06 2020 +0200

    5.0-1160

diff --git a/ripd/rip_interface.c b/ripd/rip_interface.c
index c7ab4105e..13eefdff9 100644
--- a/ripd/rip_interface.c
+++ b/ripd/rip_interface.c
@@ -1752,6 +1752,7 @@ static int rip_interface_config_write(struct vty *vty)
 		    && (ri->ri_send == RI_RIP_UNSPEC)
 		    && (ri->ri_receive == RI_RIP_UNSPEC)
 		    && (ri->auth_type != RIP_AUTH_MD5) && (!ri->v2_broadcast)
+		    && (ri->auth_type != RIP_AUTH_SIMPLE_PASSWORD)
 		    && (ri->md5_auth_len != RIP_AUTH_MD5_SIZE)
 		    && (!ri->auth_str) && (!ri->key_chain))
 			continue;
