commit a6705034bbec02e11b5e8ae1fc259025d58422e3
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 10:28:00 2020 +0200

    5.0-1170

diff --git a/configure.ac b/configure.ac
index 4d7c78271..345833bc1 100755
--- a/configure.ac
+++ b/configure.ac
@@ -1811,7 +1811,7 @@ AM_CONDITIONAL([ZEROMQ], test "x$ZEROMQ" = "xtrue")
 dnl ----------
 dnl configure date
 dnl ----------
-CONFDATE=`date '+%Y%m%d'`
+CONFDATE=0
 AC_SUBST(CONFDATE)
 
 dnl ------------------------------
