commit ea2849dfce3db6332b90b300d198b59807270afb
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 8 12:53:55 2020 +0200

    5.0-1230

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index c3d099cd8..4d7c7fd0f 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -962,6 +962,8 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp, uint8_t
 				json_object_boolean_false_add(json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf, sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
