commit d70cb076ae174c4bf8b928a2be0067702f3d8229
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:20:01 2020 +0200

    6.0-1020

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 99314ed6b..1727e6e65 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -4852,6 +4852,22 @@ static void show_ip_ospf_neighbor_detail_sub(struct vty *vty,
 		vty_out(vty, "    In the area %s via interface %s\n",
 			ospf_area_desc_string(oi->area), oi->ifp->name);
 
+	/* Show Transit Area ID */
+	if (use_json) {
+		if (memcmp(oi->ifp->name, "VLINK", 5) == 0) {
+			if (oi->vl_data != NULL) {
+				json_object_string_add(json_sub, "transitAreaId",
+					inet_ntoa(oi->vl_data->vl_area_id));
+			} else {
+				json_object_string_add(json_sub, "transitAreaId",
+					"Unknown");
+			}
+		} else {
+			json_object_string_add(json_sub, "transitAreaId",
+				ospf_area_desc_string(oi->area));
+		}
+	}
+
 	/* Show neighbor priority and state. */
 	if (use_json) {
 		json_object_int_add(json_sub, "nbrPriority", nbr->priority);
