commit 618217868a257033e6043b87de0e08a4a2b96e22
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:20:57 2020 +0200

    6.0-1050

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index d31380de9..82dcd8427 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9097,10 +9097,13 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 		if (rn->info == NULL)
 			continue;
 		int flag = 0;
+		char buf1[19];
 
 		json_route = json_object_new_object();
 		if (json) {
-			json_object_object_add(json, inet_ntoa(rn->p.u.prefix4),
+			snprintf(buf1, 19, "%s/32", inet_ntoa(rn->p.u.prefix4));
+
+			json_object_object_add(json, buf1,
 					       json_route);
 			json_object_string_add(json_route, "routeType", "R ");
 		} else {
