commit 47d1b7e0c41002ff88c55c8c596592bb7165c7af
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:22:02 2020 +0200

    6.0-1090

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 7ac538936..d7a368a9e 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9463,14 +9463,19 @@ static void show_ip_ospf_route_router(struct vty *vty, struct ospf *ospf,
 				if (or->path_type == OSPF_PATH_INTER_AREA)
 					json_object_boolean_true_add(json_route,
 								     "IA");
-				if (or->u.std.flags & ROUTER_LSA_BORDER)
-					json_object_string_add(json_route,
-							       "routerType",
-							       "abr");
-				else if (or->u.std.flags & ROUTER_LSA_EXTERNAL)
-					json_object_string_add(json_route,
-							       "routerType",
-							       "asbr");
+				if ((or->u.std.flags & (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER))
+					== (ROUTER_LSA_EXTERNAL | ROUTER_LSA_BORDER)) {
+					snprintf(buf1, sizeof(buf1), "abr asbr");
+				} else if (or->u.std.flags & ROUTER_LSA_BORDER) {
+					snprintf(buf1, sizeof(buf1), "abr");
+				} else if (or->u.std.flags & ROUTER_LSA_EXTERNAL) {
+					snprintf(buf1, sizeof(buf1), "asbr");
+				} else {
+					buf1[0] = '\0';
+				}
+				json_object_string_add(json_route,
+						       "routerType",
+						       buf1);
 			} else {
 				vty_out(vty, "%s [%d] area: %s",
 					(or->path_type == OSPF_PATH_INTER_AREA
