commit 01bc54f0bf1634933230c7b427ee4c09ccaaf013
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:22:39 2020 +0200

    6.0-1110

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 47aa5b611..0cb3c943b 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9290,8 +9290,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (or->path_type) {
@@ -9588,8 +9586,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (er->path_type) {
