commit a622de292a96b14c685c97f28bb741980d1c85fa
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:24:04 2020 +0200

    6.0-1160

diff --git a/ripd/rip_interface.c b/ripd/rip_interface.c
index 6cba8212a..db3f83855 100644
--- a/ripd/rip_interface.c
+++ b/ripd/rip_interface.c
@@ -1751,6 +1751,7 @@ static int rip_interface_config_write(struct vty *vty)
 		    && (ri->ri_send == RI_RIP_UNSPEC)
 		    && (ri->ri_receive == RI_RIP_UNSPEC)
 		    && (ri->auth_type != RIP_AUTH_MD5) && (!ri->v2_broadcast)
+		    && (ri->auth_type != RIP_AUTH_SIMPLE_PASSWORD)
 		    && (ri->md5_auth_len != RIP_AUTH_MD5_SIZE)
 		    && (!ri->auth_str) && (!ri->key_chain))
 			continue;
