commit 9c7231bde9c76fba7556054a720f3e912b1f9d0e
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:44:17 2020 +0200

    6.0-1230

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index 5a55066a8..c856cf3d8 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -955,6 +955,8 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp, uint8_t
 				json_object_boolean_false_add(json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf, sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
