commit 1b0e55f258d829e39edb381e35b0abbd4200e3cb
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 12 09:45:13 2020 +0200

    6.0-1240

diff --git a/lib/zebra.h b/lib/zebra.h
index 97b1e9531..dfa372dd3 100644
--- a/lib/zebra.h
+++ b/lib/zebra.h
@@ -53,10 +53,9 @@ typedef unsigned char uint8_t;
 #include <sys/types.h>
 #include <sys/param.h>
 #ifdef HAVE_SYS_SYSCTL_H
-#ifdef GNU_LINUX
-#include <linux/types.h>
-#endif
+#ifndef GNU_LINUX
 #include <sys/sysctl.h>
+#endif
 #endif /* HAVE_SYS_SYSCTL_H */
 #include <sys/ioctl.h>
 #ifdef HAVE_SYS_CONF_H
