diff --git a/lib/libfrr.c b/lib/libfrr.c
index 821c57f..d8378c8 100644
--- a/lib/libfrr.c
+++ b/lib/libfrr.c
@@ -879,12 +879,10 @@ out:
 	return 0;
 }
 
-void frr_run(struct thread_master *master)
+static int frr_log_version(struct thread *t)
 {
 	char instanceinfo[64] = "";
 
-	frr_vty_serv();
-
 	if (di->instance)
 		snprintf(instanceinfo, sizeof(instanceinfo), "instance %u ",
 			 di->instance);
@@ -892,6 +890,16 @@ void frr_run(struct thread_master *master)
 	zlog_notice("%s %s starting: %svty@%d%s", di->name, FRR_VERSION,
 		    instanceinfo, di->vty_port, di->startinfo);
 
+	return 0;
+}
+
+void frr_run(struct thread_master *master)
+{
+	frr_vty_serv();
+
+        // Have to defer this until a possible log file is opened.
+	thread_add_event(master, frr_log_version, NULL, 0, NULL);
+
 	if (di->terminal) {
 		vty_stdio(frr_terminal_close);
 		if (daemon_ctl_sock != -1) {
