diff --git a/zebra/kernel_netlink.c b/zebra/kernel_netlink.c
index 545b4d9..c350c32 100644
--- a/zebra/kernel_netlink.c
+++ b/zebra/kernel_netlink.c
@@ -278,6 +278,13 @@ static int netlink_information_fetch(struct nlmsghdr *h, ns_id_t ns_id,
 		return netlink_rule_change(h, ns_id, startup);
 	case RTM_DELRULE:
 		return netlink_rule_change(h, ns_id, startup);
+	case RTM_GETNEIGH:
+		// vtss_if_mux_netlink.c contains a call to rtnl_notify() with
+		// type = RTM_GETNEIGH that causes this to be notified by the
+		// kernel. I don't know why that call is in the mux, but until
+		// Allan fixes it, we must avoid the flog_err() call below,
+		// because it grows the log.
+		break;
 	default:
 		/*
 		 * If we have received this message then
