diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 5faf285..a4a5c52 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -3438,7 +3438,6 @@ show_ip_ospf_common (struct vty *vty, struct ospf *ospf, u_char use_json)
           unsigned long time_store = 0;
 
           result = tv_sub (recent_relative_time(), ospf->ts_spf);
-          result = tv_sub (result, recent_relative_time());
           time_store = (1000 * result.tv_sec) + (result.tv_usec / 1000);
           json_object_int_add(json, "spfLastExecutedMsecs", time_store);
 
