commit 052dbcb283a772edf6a3302b7c8cbc07a4158c79
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 10:39:42 2020 +0200

    7.0-1020

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 751cc8312..d486cd5c7 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -4918,6 +4918,22 @@ static void show_ip_ospf_neighbor_detail_sub(struct vty *vty,
 		vty_out(vty, "    In the area %s via interface %s\n",
 			ospf_area_desc_string(oi->area), oi->ifp->name);
 
+	/* Show Transit Area ID */
+	if (use_json) {
+		if (memcmp(oi->ifp->name, "VLINK", 5) == 0) {
+			if (oi->vl_data != NULL) {
+				json_object_string_add(json_sub, "transitAreaId",
+					inet_ntoa(oi->vl_data->vl_area_id));
+			} else {
+				json_object_string_add(json_sub, "transitAreaId",
+					"Unknown");
+			}
+		} else {
+			json_object_string_add(json_sub, "transitAreaId",
+				ospf_area_desc_string(oi->area));
+		}
+	}
+
 	/* Show neighbor priority and state. */
 	if (use_json) {
 		json_object_int_add(json_sub, "nbrPriority", nbr->priority);
