commit 307994ef65d301004d598c64d3873af5e6c9e313
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 11:09:57 2020 +0200

    7.0-1110

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index d735ca1c7..d2a2e2761 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -9405,8 +9405,6 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (or->path_type) {
@@ -9703,8 +9701,6 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 			json_route_array = json_object_new_array();
 			json_route = json_object_new_object();
 			json_object_object_add(json, buf1, json_route_array);
-			json_object_to_json_string_ext(
-				json, JSON_C_TO_STRING_NOSLASHESCAPE);
 		}
 
 		switch (er->path_type) {
