commit 9d1ed3aebe4494aa1a3e89d205594fe5229ee82d
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 13:30:07 2020 +0200

    7.0-1230

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index 1c06fadab..6ae29ddd3 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -986,6 +986,8 @@ static int ospf6_interface_show(struct vty *vty, struct interface *ifp, bool
 				json_object_boolean_false_add(json, "mtuMismatchDetect");
 			}
 			json_object_string_add(json, "area", strbuf);
+			inet_ntop(AF_INET, &oi->area->ospf6->router_id, strbuf, sizeof(strbuf));
+			json_object_string_add(json, "router_id", strbuf);
 			json_object_int_add(json, "cost", oi->cost);
 		} else {
 			vty_out(vty,
