commit d5484760413b58bdb2811f2345d7d898f920214d
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 13:30:24 2020 +0200

    7.0-1240

diff --git a/lib/zebra.h b/lib/zebra.h
index 09115951e..87d61392b 100644
--- a/lib/zebra.h
+++ b/lib/zebra.h
@@ -52,10 +52,9 @@ typedef unsigned char uint8_t;
 #include <sys/types.h>
 #include <sys/param.h>
 #ifdef HAVE_SYS_SYSCTL_H
-#ifdef GNU_LINUX
-#include <linux/types.h>
-#endif
+#ifndef GNU_LINUX
 #include <sys/sysctl.h>
+#endif
 #endif /* HAVE_SYS_SYSCTL_H */
 #include <sys/ioctl.h>
 #ifdef HAVE_SYS_CONF_H
