commit 140329889e301c6746810a53bb365f8a2a4c0b5e
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 13:35:17 2020 +0200

    7.0-1270

diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index d2a2e2761..132d9de4b 100644
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -53,6 +53,17 @@
 #include "ospfd/ospf_dump.h"
 #include "ospfd/ospf_bfd.h"
 
+#ifndef FALSE
+// Used to be defined in json_object.h (from libjson-c), but was removed with
+// the release of json-c-0.14-20200419.
+#define FALSE 0
+#endif
+#ifndef TRUE
+// Used to be defined in json_object.h (from libjson-c), but was removed with
+// the release of json-c-0.14-20200419.
+#define TRUE 1
+#endif
+
 static const char *ospf_network_type_str[] = {
 	"Null",	"POINTOPOINT", "BROADCAST", "NBMA", "POINTOMULTIPOINT",
 	"VIRTUALLINK", "LOOPBACK"};
