commit b7a16ca762c338c2c045bf4f34ca98352446c1c3
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Thu Oct 22 13:35:59 2020 +0200

    7.0-1280

diff --git a/zebra/kernel_netlink.c b/zebra/kernel_netlink.c
index 360f596b8..69ec1b1a5 100644
--- a/zebra/kernel_netlink.c
+++ b/zebra/kernel_netlink.c
@@ -279,6 +279,13 @@ static int netlink_information_fetch(struct nlmsghdr *h, ns_id_t ns_id,
 		return netlink_rule_change(h, ns_id, startup);
 	case RTM_DELRULE:
 		return netlink_rule_change(h, ns_id, startup);
+	case RTM_GETNEIGH:
+		// vtss_if_mux_netlink.c contains a call to rtnl_notify() with
+		// type = RTM_GETNEIGH that causes this to be notified by the
+		// kernel. I don't know why that call is in the mux, but until
+		// Allan fixes it, we must avoid the flog_err() call below,
+		// because it grows the log.
+		break;
 	default:
 		/*
 		 * If we have received this message then
