commit 8879d4e419eb0d78c3d373d977d6e769f80a9bd0
Author: Rene Nielsen <rene.nielsen@microchip.com>
Date:   Mon Oct 26 15:59:37 2020 +0100

    RBN

diff --git a/configure.ac b/configure.ac
index 97559ac..69dc0d2 100755
--- a/configure.ac
+++ b/configure.ac
@@ -1613,19 +1613,8 @@ CFLAGS="$ac_cflags_save"
 
 ac_libs_save="$LIBS"
 LIBS="$LIBS $LIBYANG_LIBS"
-AC_CHECK_FUNC([ly_register_types], [
   libyang_ext_builtin=true
   AC_DEFINE([LIBYANG_EXT_BUILTIN], [1], [have ly_register_types()])
-], [
-  libyang_ext_builtin=false
-  AC_MSG_WARN([===== old libyang (before 0.16.74) detected =====])
-  AC_MSG_WARN([The available version of libyang does not seem to support])
-  AC_MSG_WARN([built-in YANG extension modules.  This will cause "make check"])
-  AC_MSG_WARN([to fail and may create installation and version mismatch issues.])
-  AC_MSG_WARN([Support for the old mechanism will be removed at some point.])
-  AC_MSG_WARN([Please update libyang to version 0.16.74 or newer.])
-  AC_MSG_WARN([===== old libyang (before 0.16.74) detected =====])
-])
 AM_CONDITIONAL([LIBYANG_EXT_BUILTIN], [$libyang_ext_builtin])
 LIBS="$ac_libs_save"
 
